crates/ide-assists/src/handlers/convert_nested_function_to_closure.rs

```diff
@@ -0,0 +1,147 @@
+use hir::Semantics;
+use ide_db::assists::{AssistId, AssistKind};
+use ide_db::base_db::FileId;
+use ide_db::RootDatabase;
+use syntax::ast::{self, HasName};
+use syntax::{AstNode, SyntaxKind};
+
+use crate::assist_context::{AssistContext, Assists};
+
+// Assist: convert_nested_function_to_closure
+//
+// Converts a function that is defined within the body of another function into a closure.
+//
+// ```
+// fn main() {
+//     fn foo(label: &str, $0number: u64) {
+//         println!("{}: {}", label, number);
+//     }
+//
+//     foo("Bar", 100);
+// }
+// ```
+// ->
+// ```
+// fn main() {
+//     let foo = |label: &str, number: u64| {
+//         println!("{}: {}", label, number);
+//     };
+//
+//     foo("Bar", 100);
+// }
+// ```
+pub(crate) fn convert_nested_function_to_closure(
+    acc: &mut Assists,
+    ctx: &AssistContext<'_>,
+) -> Option<()> {
+    let function = ctx.find_node_at_offset::<ast::Fn>()?;
```

Comment:
It would probably be better if we only trigger the assist when we are on the function's name, so
```rs
    let name = ctx.find_node_at_offset::<ast::Name>()?;
    let function = name.syntax().parent().and_then(ast::Fn::cast)?;
```